README
This is the project page for my final research project for CSCI 198. 
I will be using the PyBrain library for constructing a neural network.
The neural network will be used to predict future stock price given various inputs.

INSTALLATION
============

Unix-like OSes (i.e. Linux, Mac OS X)
=====================================

Installation instructions with easy_install:
1) Install pybrain with easy_install pybrain. 
2) Run aifinalproject.py with Python >= 2.7.x

OR

1) Install setuptools/easy_install 
2) Install SciPY and matplotlib with easy_install: (i.e. easy_install scipy && easy_install matplotlib)
3) Navigate to pybrain folder, then follow instructions from there. 

NOTE: I have only tested this on Debian 6.0 x86. For Linux users, it might be better to use your distribution's package manager to install scipy and matplotlib.

Windows
=======
I have not tested this on Windows. Try following the above instructions. 


CHANGELOG
05-06-2012 Naveen Akkapeddi <naveen.csci@gmail.com>
	* Submitting final project. While the results weren't particulary 
	robust, with a proper dataset this project can be trivially modified
	to produce good results that can be used in a trading strategy.
	* aifinalproject.py Fixed the main program from halting by changing 
	the training parameters. The network is trained for 100 epochs, 5 times 
	per epoch. Made the backprop trainer verbose to display total error
	during the training sequence.

04-24-2012 Naveen Akkapeddi <naveen.csci@gmail.com>
	* aifinalproject.py The training dataset will now be automatically
	generated. Converted each item in nested list to float, then natural log
	is taken for each item. 
	Started building the network with 2 input neurons, 2 hidden neurons, and 1 output neuron.

04-23-2012 Naveen Akkapeddi <naveen.csci@gmail.com>

	* datadumper.py: Created module which will take data from Yahoo! finance 
	and pickle it into a file.
	* aifinalproject.py Created main file that will be used to construct 
	the neural network.
	* ystockquote.py: Fixed an issue where date was being incorrectly passed to the Yahoo! finance API. 

04-19-2012 Naveen Akkapeddi <naveen.csci@gmail.com>

	* CHANGELOG: Added changelog to track changes to the project.
	* ystockquote.py: Added ystockquote library for python interface to
	Yahoo! finance. Fixed a TypeError issue present in the original 
	library by creating four local variables that represented ints as
	strings. 

04-18-2012 Naveen Akkapeddi <naveen.csci@gmail.com>
	* README: Added a readme describing the project.


aifinalproject.py
#!/usr/bin/env python
#
#  Copyright (c) 2012, Naveen Akkapeddi (naveen.csci@gmail.com)
#
#  license: GNU LGPL
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 2.1 of the License, or (at your option) any later version.
#  

import sys
import datadumper
import pickle
import math
from pybrain.tools.shortcuts import buildNetwork
from pybrain.structure import TanhLayer, SoftmaxLayer
from pybrain.datasets import SupervisedDataSet
from pybrain.supervised.trainers import BackpropTrainer
from pybrain.utilities import percentError

print """Generating dataset (historical stock data)..."""
datadumper.data_dump()

myList = [] 	# Create empty list to store nested list containing closing price and volume
stockFile = open('historicaldata.pickle', 'r')
while 1:
    try:
        myList.append(pickle.load(stockFile))
    except (EOFError):
        break
stockFile.close()

for index1, row in enumerate(myList): # For each item in every row of the nested list, convert string to float and apply natural log function
	for index2, item in enumerate(row):
		try:
			myList[index1][index2] = int(item)
		except ValueError:
			pass
#print myList

myClonedList = myList # Alias the previous nested list, take the first item (natural log closing price) of every row and append it to new list.
targetData = []
for index1, row in enumerate(myClonedList):
	targetData.append(myClonedList[index1][0])

targetData.pop() #Pop off last element because it should be the next day's closing price.
#print targetData

myNet = buildNetwork(2, 2, 1, hiddenclass=SoftmaxLayer) # Build network with 2 input neurons, 2 hidden neurons, and 1 output neuron
print myNet['in'], myNet['hidden0'], myNet['out'] # Debug message to confirm network setup

myDS = SupervisedDataSet(2, 1) # Create dataset with two dimensional input, one dimensional target.

for i in enumerate(myList): #Pop off each element, then added to dataset
	myDataSetAdder = myDS.appendLinked(myList.pop(), targetData.pop())

for inpt, target in myDS: #display dataset structure
	print inpt, target

myTrainer = BackpropTrainer(myNet, myDS, verbose=True)
print """This may take awhile..."""
myTrainData, myTestData = myDS.splitWithProportion(0.25)
print "Number of training patterns: ", len(myTestData)
print "Input and output dimensions: ", myTestData.indim, myTrainData.outdim
print "First sample (input, target):"
print myTrainData['input'][0], myTrainData['target'][0]
#import pdb; pdb.set_trace() # Debugger, uncomment to run python debugger

for i in range(20):
	myTrainer.trainEpochs( 5 ) # Run the network for 5 epochs...
	trnresult = percentError (myTestData, myTrainData)
    	print "epoch: %4d" % myTrainer.totalepochs, \
        "  train error: %5.2f%%" % trnresult, \
#        "  test error: %5.2f%%" % tstresult


ystockquote.py
#!/usr/bin/env python
#
#  Copyright (c) 2007-2008, Corey Goldberg (corey@goldb.org)
#
#  license: GNU LGPL
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 2.1 of the License, or (at your option) any later version.


import urllib


"""
This is the "ystockquote" module.

This module provides a Python API for retrieving stock data from Yahoo Finance.

sample usage:
>>> import ystockquote
>>> print ystockquote.get_price('GOOG')
529.46
"""


def __request(symbol, stat):
    url = 'http://finance.yahoo.com/d/quotes.csv?s=%s&f=%s' % (symbol, stat)
    return urllib.urlopen(url).read().strip().strip('"')


def get_all(symbol):
    """
    Get all available quote data for the given ticker symbol.
    
    Returns a dictionary.
    """
    values = __request(symbol, 'l1c1va2xj1b4j4dyekjm3m4rr5p5p6s7').split(',')
    data = {}
    data['price'] = values[0]
    data['change'] = values[1]
    data['volume'] = values[2]
    data['avg_daily_volume'] = values[3]
    data['stock_exchange'] = values[4]
    data['market_cap'] = values[5]
    data['book_value'] = values[6]
    data['ebitda'] = values[7]
    data['dividend_per_share'] = values[8]
    data['dividend_yield'] = values[9]
    data['earnings_per_share'] = values[10]
    data['52_week_high'] = values[11]
    data['52_week_low'] = values[12]
    data['50day_moving_avg'] = values[13]
    data['200day_moving_avg'] = values[14]
    data['price_earnings_ratio'] = values[15]
    data['price_earnings_growth_ratio'] = values[16]
    data['price_sales_ratio'] = values[17]
    data['price_book_ratio'] = values[18]
    data['short_ratio'] = values[19]
    return data
    
    
def get_price(symbol): 
    return __request(symbol, 'l1')


def get_change(symbol):
    return __request(symbol, 'c1')
    
    
def get_volume(symbol): 
    return __request(symbol, 'v')


def get_avg_daily_volume(symbol): 
    return __request(symbol, 'a2')
    
    
def get_stock_exchange(symbol): 
    return __request(symbol, 'x')
    
    
def get_market_cap(symbol):
    return __request(symbol, 'j1')
   
   
def get_book_value(symbol):
    return __request(symbol, 'b4')


def get_ebitda(symbol): 
    return __request(symbol, 'j4')
    
    
def get_dividend_per_share(symbol):
    return __request(symbol, 'd')


def get_dividend_yield(symbol): 
    return __request(symbol, 'y')
    
    
def get_earnings_per_share(symbol): 
    return __request(symbol, 'e')


def get_52_week_high(symbol): 
    return __request(symbol, 'k')
    
    
def get_52_week_low(symbol): 
    return __request(symbol, 'j')


def get_50day_moving_avg(symbol): 
    return __request(symbol, 'm3')
    
    
def get_200day_moving_avg(symbol): 
    return __request(symbol, 'm4')
    
    
def get_price_earnings_ratio(symbol): 
    return __request(symbol, 'r')


def get_price_earnings_growth_ratio(symbol): 
    return __request(symbol, 'r5')


def get_price_sales_ratio(symbol): 
    return __request(symbol, 'p5')
    
    
def get_price_book_ratio(symbol): 
    return __request(symbol, 'p6')
       
       
def get_short_ratio(symbol): 
    return __request(symbol, 's7')
    
    
def get_historical_prices(symbol, start_date, end_date):
    """
    Get historical prices for the given ticker symbol.
    Date format is 'YYYYMMDD'
    
    Returns a nested list.
    """

    str_end_date = str(int(end_date)) #create local variables to convert int date to string date (nakkapeddi) 
    str_start_date = str(int(start_date))
    str_end_date_minus_one = str(int(end_date-100))
    str_start_date_minus_one = str(int(start_date-100))

    url = 'http://ichart.yahoo.com/table.csv?s=%s&' % symbol + \
          'd=%s&' % str_end_date_minus_one[4:6] + \
          'e=%s&' % str_end_date[6:8] + \
          'f=%s&' % str_end_date[0:4] + \
          'g=d&' + \
          'a=%s&' % str_start_date_minus_one[4:6] + \
          'b=%s&' % str_start_date[6:8] + \
          'c=%s&' % str_start_date[0:4] + \
          'ignore=.csv'
    days = urllib.urlopen(url).readlines()
    data = [day[:-2].split(',') for day in days]
    return data
        
